source/model/backtraceMetrics.ts

```diff
@@ -0,0 +1,232 @@
+import { BacktraceClientOptions } from '..';
+import {
+  currentTimestamp,
+  getBacktraceGUID,
+  getEndpointParams,
+  post,
+  uuid,
+} from '../utils';
+import { APP_NAME, USER_AGENT, VERSION } from '../consts/application';
+import { SEC_TO_MILLIS } from '../consts';
+
+/**
+ * Handles Backtrace Metrics.
+ */
+export class BacktraceMetrics {
+  private readonly universe: string;
+  private readonly token: string;
+  private readonly hostname: string;
+
+  private readonly timeout: number = 15000; // Fifteen seconds in milliseconds.
+  private readonly persistenceInterval: number = 1800000; // Thirty minutes in milliseconds.
+  private readonly heartbeatInterval: number = 60000; // One minutes in milliseconds.
+
+  private readonly timestamp = currentTimestamp();
+  private readonly userAgent = USER_AGENT;
+  private readonly applicationName = APP_NAME;
+  private readonly applicationVersion = VERSION;
+
+  private summedEndpoint: string;
+  private uniqueEndpoint: string;
+
+  private readonly guid;
+  private sessionId;
+  private lastActive;
+
+  constructor(
+    configuration: BacktraceClientOptions,
+    hostname = 'https://events.backtrace.io',
+  ) {
+    if (!configuration.endpoint) {
+      throw new Error(`Backtrace: missing 'endpoint' option.`);
+    }
+    const { universe, token } = getEndpointParams(configuration.endpoint);
+
+    if (!universe) {
+      throw new Error(
+        `Backtrace: 'universe' could not be parsed from the endpoint.`,
+      );
+    }
+
+    if (!token && !configuration.token) {
+      throw new Error(
+        `Backtrace: missing 'token' option or it could not be parsed from the endpoint.`,
+      );
+    }
+
+    this.universe = universe;
+    this.token = configuration.token || token || ''; // default '' case needed since type checking isn't certain that type can not be undefined.
+    this.timeout = configuration.timeout;
+    this.hostname = hostname;
+
+    this.summedEndpoint = `${this.hostname}/api/unique-events/submit?universe=${this.universe}&token=${this.token}`;
+    this.uniqueEndpoint = `${this.hostname}/api/summed-events/submit?universe=${this.universe}&token=${this.token}`;
+
+    this.guid = getBacktraceGUID();
+    this.sessionId = this.getSessionId();
+    this.lastActive = this.getLastActive();
+
+    this.persistSession(); // Create/persist session on construction.
+
+    // Get current session interval. If one is set or running, clear it.
+    const currentIntervalId = this.getActiveSessionIntervalId();
+    if (currentIntervalId) {
+      clearInterval(currentIntervalId);
+    }
+    // Start new interval.
+    // Persist session if page is focused on heartbeat interval
+    const intervalId = setInterval(
+      () => this.persistIfFocused(),
+      this.heartbeatInterval,
```

Comment:
one more thing that I think we should consider here. If someone keeps refreshing the website, we will never send a next unique/summed events, because `heartbeatInterval` will stay the same, Should we decrease it based on some kind of information from local storage when we should send next data ?

Reply:
When the page is refreshed, we first persist the session and then start the heartbeat. So it will be persisted fine in that case